Core/include/Acts/Geometry/Portal.hpp

```diff
@@ -0,0 +1,234 @@
+#include "Acts/Utilities/Result.hpp"
+
+#include <memory>
+
+namespace Acts {
+
+class RegularSurface;
+class GeometryContext;
+class TrackingVolume;
+class CylinderSurface;
+class PlaneSurface;
+class DiscSurface;
+class Surface;
+
+class PortalLinkBase;
+
+/// Exception thrown when portals cannot be merged
+class PortalMergingException : public std::exception {
+  const char* what() const noexcept override {
+    return "Failure to merge portals";
+  }
+};
+
+/// Exception thrown when portals cannot be fused
+class PortalFusingException : public std::exception {
+  const char* what() const noexcept override {
+    return "Failure to fuse portals";
+  }
+};
+
+/// A portal connects two or more neighboring volumes. Each volume has a set of
+/// portals that describes which volumes lie behind the portal in that
+/// direction. Portals use associated portal links to perform lookups of target
+/// volumes.
+/// Each portal has two links, and a corresponding surface. One link is
+/// associated with the direction along the surface's normal vector, and one
+/// with the opposite direction.
+class Portal {
+ public:
+  /// Constructor for a portal from a single link
+  /// @param direction The direction of the link
+  /// @param link The portal link
+  Portal(Direction direction, std::unique_ptr<PortalLinkBase> link);
+
+  /// Constructor for a portal from a surface and volume, where a trivial portal
+  /// link is automatically constructed.
+  /// @param direction The direction of the link
+  /// @param surface The surface from which to create the portal link
+  /// @param volume The volume this portal connects to in the @p direction
+  ///               relative to the normal of @p surface.
+  Portal(Direction direction, std::shared_ptr<RegularSurface> surface,
+         TrackingVolume& volume);
+
+  /// Constructor for a portal from two links. One of the links can be
+  /// `nullptr`, but at least one of them needs to be set. If both are set, they
+  /// need to be valid compatible links that can be fused.
+  /// @param gctx The geometry context
+  /// @param alongNormal The link along the normal of the surface
+  /// @param oppositeNormal The link opposite to the normal of the
+  Portal(const GeometryContext& gctx,
+         std::unique_ptr<PortalLinkBase> alongNormal,
+         std::unique_ptr<PortalLinkBase> oppositeNormal);
+
+  /// Helper struct for the arguments to the portal constructor below using
+  /// designated initializers.
+  struct Arguments {
+    /// Aggregate over a surface and a volume with optional semantics
+    struct Link {
+      Link() = default;
+      /// Constructor from a surface and a volume
+      Link(std::shared_ptr<RegularSurface> surface, TrackingVolume& volume)
+          : m_surface(std::move(surface)), m_volume(&volume) {}
+
+      /// The associated surface
+      std::shared_ptr<RegularSurface> m_surface = nullptr;
+      /// The associated volume
+      TrackingVolume* m_volume = nullptr;
```

Comment:
nitpick: for plain public struct members we usually do `surface` instead of `m_surface`